File: dsp_2026_2/04_error_cuantizacion.py
#!/usr/bin/env python3
"""04_error_cuantizacion.py

Tema 1 - Error de cuantización: modelo de ruido uniforme.
Ejecutar:  python3 04_error_cuantizacion.py

Uso de SciPy: scipy.stats.uniform para comparar con la distribución teórica.
"""

import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import uniform

Vmax, Vmin = 1.0, -1.0
b = 8
Delta = (Vmax - Vmin) / (2**b)

rng = np.random.default_rng(0)
N = 20000
x = 0.9 * (2 * rng.random(N) - 1)       # señal uniforme en [-0.9, 0.9]
xq = Delta * np.round(x / Delta)
e = x - xq

fig, ax = plt.subplots()
ax.hist(e, bins=50, density=True, alpha=0.7, label="Error empírico")
xpdf = np.linspace(-Delta, Delta, 200)
ax.plot(xpdf, uniform.pdf(xpdf, loc=-Delta / 2, scale=Delta), "r--",
        label=r"Uniforme en $[-\Delta/2, \Delta/2]$")
ax.set_title("Histograma del error de cuantización")
ax.set_xlabel("e[n]")
ax.set_ylabel("Densidad")
ax.legend()
plt.tight_layout()

rms_e = np.sqrt(np.mean(e**2))
print(f"b = {b} bits, Delta = {Delta:.5f}")
print(f"RMS empírico del error: {rms_e:.6f}")
print(f"Teórico Delta/sqrt(12): {Delta / np.sqrt(12):.6f}")

# Ejercicio 1: repita con b = 4 y b = 12; compare el RMS empírico con
# Delta/sqrt(12) y verifique que el error queda acotado entre -Delta/2 y
# Delta/2.
# Ejercicio 2 (propuesto): use un seno a plena escala (x = 0.9*cos(2*pi*2*t))
# en vez de uniforme. El histograma del error deja de ser plano: explique por
# qué (la densidad de muestras no es uniforme cerca de los extremos del seno).

plt.show()

File: dsp_2026_2/04_error_cuantizacion.m
% 04_error_cuantizacion.m
% Tema 1 - Error de cuantizacion: modelo de ruido uniforme.
% Ejecutar:  octave-cli 04_error_cuantizacion.m
%

clear; close all; clc;

Vmax = 1; Vmin = -1;
b = 8;
Delta = (Vmax - Vmin) / (2^b);

N = 20000;
rng(0);
x = 0.9 * (2*rand(1, N) - 1);        % senal uniforme en [-0.9, 0.9]
xq = Delta * round(x / Delta);
e = x - xq;

figure('Name', 'Error de cuantizacion');
hist(e, 50);
hold on;
line([-Delta/2 -Delta/2], ylim, 'Color', 'r', 'LineStyle', '--');
line([ Delta/2  Delta/2], ylim, 'Color', 'r', 'LineStyle', '--');
title('Histograma del error de cuantizacion');
xlabel('e[n]'); ylabel('Conteo');

rms_e = sqrt(mean(e.^2));
fprintf('b = %d bits, Delta = %.5f\n', b, Delta);
fprintf('RMS empirico del error:    %.6f\n', rms_e);
fprintf('Teorico Delta/sqrt(12):    %.6f\n', Delta/sqrt(12));

%% Ejercicio 1
% Repita con b = 4 y b = 12. Compare el RMS empirico con Delta/sqrt(12) en
% cada caso y verifique que el error queda acotado entre -Delta/2 y Delta/2.

%% Ejercicio 2 (propuesto)
% Use una senal sinusoidal a plena escala (x = 0.9*cos(2*pi*2*t)) en vez de
% uniforme. El histograma del error deja de ser plano: explique por que
% (la densidad de muestras no es uniforme cerca de los extremos del seno).
